url5a01f2c1d9_core9_core_tgz-pJ8-io_core9_core_boot_BootstrapFrameworkJ8/src/io/core9/core/proxy/DefaultInvocationHandler.java

```diff
@@ -0,0 +1,61 @@
+package io.core9.core.proxy;
+
+import io.core9.core.hooks.Hook;
+import io.core9.core.hooks.HookProvider;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.util.Collection;
+
+public class DefaultInvocationHandler implements InvocationHandler {
+	private Object target;
+	private HookProvider hookprovider = HookProvider.getInstance();
+
+	public DefaultInvocationHandler(Object target) {
+		this.target = target;
+	}
+
+	@Override
+	public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+		String name = proxy.getClass().getInterfaces()[0].getName();
+		Collection<Hook> pres = hookprovider.getHooks(name, method.getName() + ":pre");
+		Collection<Hook> posts = hookprovider.getHooks(name, method.getName() + ":post");
+		
+		Object[] newargs = null;
+		if(pres != null) {
+			for(Hook hook : pres) {
+				newargs = hook.execute(args);
+			}
+			if(newargs != null) {
+				args = newargs;
+			}
+		}
+		Object ret = execute(proxy, method, args);
+		if(posts != null) {
+			for(Hook hook : posts) {
+				if(ret == null) {
+					ret = hook.execute(args);
+				} else {
+					ret = hook.execute(ret)[0];
+				}
+			}
+		}
+		return ret;
+	}
+	
+	/**
+	 * Executes the real called method (can be overridden)
+	 * @param proxy
+	 * @param method
+	 * @param args
+	 * @return Object
+	 * @throws Throwable
+	 */
+	protected Object execute(Object proxy, Method method, Object[] args) throws Throwable {
+		return method.invoke(target, args);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)